haskell/DemystifyingFoldable.hs

```diff
@@ -0,0 +1,45 @@
+{-
+
+Foldable represents a structure that, given a suitable "folding" function, can
+be collapsed into a new value without the structure.
+
+In these reps you'll practice the various ways Foldable can be implemented and
+the how to implement it on different kinds of datastructures.
+
+- Implement a type like Maybe
+- Write a full instance of Foldable for your "Maybe"
+- Write a newtype around List
+- Write a full instance of Foldable for your "List"
+
+-}
+
+
+module DemystifyingFoldable where
+
+data Option a
+  = Something a
+  | None
+
+instance Foldable Option where
+  foldr fn seed opt =
+    case opt of
+      Something a -> fn a seed
+      None        -> seed
+
+  foldMap fn opt =
+    case opt of
+      None        -> mempty
+      Something a -> fn a
+
+newtype Listy a = Listy [a]
+
+instance Foldable Listy where
+  foldr fn seed lst =
+    case lst of
+      Listy [] -> seed
+      Listy (first : rest) -> foldr fn (fn first seed) (Listy rest)
```

Comment:
Oops... this implementation in `foldl` rather than `foldr`, right?

Reply:
:man_facepalming:   I've been doing my reps with single-item data-structures for so long that I wasn't even considering the difference.